pylinalg/func/vector.py

```diff
@@ -116,3 +116,114 @@ def vector_apply_matrix(vectors, matrix, /, *, w=1, out=None, dtype=None):
     if dtype is not None:
         out = out.astype(dtype, copy=False)
     return out
+
+
+def project_camera(vector, projection_matrix) -> np.ndarray:
+    """Project a 3D vector into 2D space.
+
+    Parameters
+    ----------
+    vector : ArrayLike
+        The vector to be projected.
+    projection_matrix: ArrayLike
+        The camera's intrinsic matrix.
+
+    Returns
+    -------
+    projected_vector : ndarray
+        The projection of the vector.
+
+    """
+
+    raise NotImplementedError()
+
+
+def unproject_camera(vector, projection_matrix, /, *, depth=0) -> np.ndarray:
+    """Un-project a vector from 2D space to 3D space.
+
+    Parameters
+    ----------
+    vector : ArrayLike
+        The vector to be un-projected.
+    projection_matrix: ArrayLike
+        The camera's intrinsic matrix.
+    depth : float
+        The distance of the unprojected vector from the camera.
+
+    Returns
+    -------
+    projected_vector : ndarray
+        The unprojected vector in 3D space
+    """
+
+    raise NotImplementedError()
+
+
+def apply_quaternion(vector, quaternion) -> np.ndarray:
+    """Rotate a vector using a quaternion.
+
+    Parameters
+    ----------
+    vector : ArrayLike
+        The vector to be rotated.
+    quaternion : ArrayLike
+        The quaternion to apply (in xyzw format).
+
+    Returns
+    -------
+    rotated_vector : np.ndarray
+        The rotated vector.
+
+    """
+
+    raise NotImplementedError()
+
+
+def from_spherical(spherical) -> np.ndarray:
+    """Convert Spherical --> Euclidian coordinates
+
+    Parameters
+    ----------
+    spherical : ArrayLike
+        A vector in spherical coordinates (r, phi, theta).
+
+    Returns
+    -------
+    euclidean_coordinates : ArrayLike
+        A vector in euclidian coordinates.
+
+    """
+
+    raise NotImplementedError()
+
+
+def distance_to(vectorA, vectorB, /) -> np.ndarray:
+    """The distance between two vectors
+
+    Parameters
+    ----------
+    vectorA : ArrayLike
+        The first vector.
+    vectorB : ArrayLike
+        The second vector.
+
+    Returns
+    -------
+    distance : np.ndarray
+        The distance between both vectors.
+
+    """
+
+    raise NotImplementedError()
+
+
+def from_matrix_position(homogeneous_matrix) -> np.ndarray:
+    """Return the position of the matrix (??)
+
+    @Korijn: I can't work out what this function does on a high-level. It
+    clearly extracts the bottom row of a homogeneous matrix, but why do we want
+    that?
```

Comment:
it just standardizes the conventions; where in the matrix is the translation component stored? this depends on whether the library is using row or column major matrix representations.

Reply:
oh. I've never seen it stored in the bottom row nor considered it. Is that for performance reasons or pure convention?

Reply:
Um, could be either or both reasons! Depends on the data you're working with.